### DISCUSSION TITLE
Using the Graph API from a Power App

### DISCUSSION THREAD

**Discussion by greddin** [microsoft/hack-together#51] [Q&A] [answered]
Hi all, I am creating a project that uses a Power App to call the Microsoft Graph API. The idea is to return only those users who have an Employee Type of Contractor. 

Using the Graph Explorer I'm able to use the following query that returns only users which are Contractors:
`https://graph.microsoft.com/v1.0/users?$filter=employeeType eq 'Contractor'&$count=true`

I have registered an app in the Azure Active Directory and created a custom connector with one Action called GetUsers for my Power App.

For testing, in the Power App I just have a button and trying to use the following command on the OnSelect.

`ClearCollect(Contractors, 'graph-user-contractors'.GetUsers({employeeType eq 'Contractor',true,'eventual'}).value)`

I'm getting an error in the Power App contextual help that is saying there are some invalid arguments. I'm not sure how to pass in the parameters syntax in for the ClearCollect. 

![2023-03-02_141103](https://user-images.githubusercontent.com/967815/222542574-5bd44eb7-86ec-42ba-aba1-7f07add77135.png)

After created the custom connector it seems there are three parameters required.

$filter
$count (true/false) - it seems this is required when filtering on the employeeType
ConsistencyLevel - it seems this is required when using the filter as well.  Setting it to 'eventual' works in the Graph Explorer.

Thank you for any help you can provide or point me in the right direction to!

**Comment by waldekmastykarz** [comment]
As you're building a Power App, I assume your work is not related to this hack because it's not using .NET. Instead you're submitting a generic question, correct? Let me ask around for you if anyone knows what could be wrong.

**Comment by LuiseFreese** [comment]
Heya @greddin, happy to help :-)  Few questions: 

1. Did you test your custom connector in the Custom Connector Wizard experience?
2. Did you add your custom connector to your app?
3. Can you share the definition and the test tab of the custom connector and the query in the Graph Explorer as a screenshot?

**Reply by greddin** [reply]
Hi Luise, yes the test for the custom connector works and I can also use the custom connect in a Power Automate flow without issue. I'm afraid I just don't know the syntax good enough to call it from the Power App. 

I've attached the test and definition tabs for the custom connector. 

![custom-connector-test-tab](https://user-images.githubusercontent.com/967815/222739337-c6ff70d6-0c64-40ce-aec7-de5c975cf912.png)

![custom-connector-definition-tab](https://user-images.githubusercontent.com/967815/222739366-9edb5507-80f4-4e07-8a8c-e2a096b333df.png)

Btw - I enjoyed watching your video on creating the custom Teams meeting planner and also your article on the graph look-a-like people picker.

**Reply by LuiseFreese** [reply]
That’s already a great start. In Power Apps, you need to replace the `eq` by an `=` - you can't have any spaces in  there. Thanks for your kind feedback! Let me know how it  goes!

**Reply by greddin** [reply]
Here's my attempt now.  Do I need to enter the parameter names as well? 

![powerapp-error](https://user-images.githubusercontent.com/967815/222754674-c17439ae-6dc8-4cd0-838c-2de7fd4fff5c.png)

**Comment by LuiseFreese** [comment]
If you like to, ping me on Twitter @luisefreese or  LinkedIn https://linkedin.com/in/luisefreese and then I can give you my email so that I could jump on a quick Teams call with you.

Luise Freese




Here's my attempt now. Do I need to enter the parameter names as well?

[powerapp-error]<https://user-images.githubusercontent.com/967815/222754674-c17439ae-6dc8-4cd0-838c-2de7fd4fff5c.png>

**Comment by greddin** [comment, marked as answer]
Thanks to @LuiseFreese for helping solve my issue! 

My main issue was that the employeeType was not getting returned in the response. So the Power App would error.  We simplified the Graph query in the Explorer to: `https://graph.microsoft.com/v1.0/users?$select=employeeType,id,displayName`

We then updated the Collection in the Power App to Filter on employeeType="Contractor".
`ClearCollect(Contractors, Filter('graph-user-contractors'.GetUsers({'$select':"employeeType,id,displayName"}).value,employeeType="Contractor"))`

Thank you again @LuiseFreese for all your help!

-Greg

**Reply by waldekmastykarz** [reply]
Awesome! Great to hear you're unblocked and thank you so much @LuiseFreese for helping @greddin 👏 ❤️
